core/trino-main/src/main/java/io/trino/sql/planner/plan/TableScanNode.java

```diff
@@ -101,11 +107,38 @@ public TableScanNode(
         this.outputSymbols = ImmutableList.copyOf(requireNonNull(outputs, "outputs is null"));
         this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
         checkArgument(assignments.keySet().containsAll(outputs), "assignments does not cover all of outputs");
-        this.enforcedConstraint = requireNonNull(enforcedConstraint, "enforcedConstraint is null");
+        requireNonNull(enforcedConstraint, "enforcedConstraint is null");
+        validateEnforcedConstraint(enforcedConstraint, outputs, assignments);
+        this.enforcedConstraint = enforcedConstraint;
         this.updateTarget = updateTarget;
         this.useConnectorNodePartitioning = requireNonNull(useConnectorNodePartitioning, "useConnectorNodePartitioning is null");
     }
 
+    private static void validateEnforcedConstraint(TupleDomain<ColumnHandle> enforcedConstraint, List<Symbol> outputs, Map<Symbol, ColumnHandle> assignments)
+    {
+        verify(!enforcedConstraint.isNone(), "Table scan with NONE enforcedConstraint should be converted to ValuesNode");
```

Comment:
This is true, but there's nothing illegal about a TableScan that is guaranteed to produce no rows (i.e., enforcedConstraint == NONE). An optimizer should not be forced to do this transformation.

E.g., it should be ok for an rule to push something into a connector that results in constraint == NONE and another rule to transform a TableScan with constraint == NONE => ValuesNode.

Reply:
It's subjective. We can decide it's OK to create TableScan with enforcedConstraint=NONE or we can decide it is not OK.
Note the addition here is in response to an already codified assumption in a rule implementation, not my own invention now.
I will drop the commit for now still.